Samples/VirtualCamera/VirtualCameraMediaSource/HWMediaSource.cpp

```diff
@@ -0,0 +1,626 @@
+        *ppAttributes = nullptr;
+
+        RETURN_IF_FAILED(_CheckShutdownRequiresLock());
+        wil::com_ptr_nothrow<IMFMediaSourceEx> spMediaSourceEx;
+        RETURN_IF_FAILED(m_spDevSource->QueryInterface(IID_PPV_ARGS(&spMediaSourceEx)));
+        RETURN_IF_FAILED(spMediaSourceEx->GetStreamAttributes(dwStreamIdentifier, ppAttributes));
+
+        return S_OK;
+    }
+
+    IFACEMETHODIMP HWMediaSource::SetD3DManager(
+            _In_opt_ IUnknown* pManager
+        )
+    {
+        // Return code is ignored by the frame work, this is a
+        // best effort attempt to inform the media source of the
+        // DXGI manager to use if DX surface support is available.
+        winrt::slim_lock_guard lock(m_Lock);
+
+        RETURN_IF_FAILED(_CheckShutdownRequiresLock());
+
+        wil::com_ptr_nothrow<IMFMediaSourceEx> spMediaSourceEx;
+        RETURN_IF_FAILED(m_spDevSource->QueryInterface(IID_PPV_ARGS(&spMediaSourceEx)));
+        RETURN_IF_FAILED(spMediaSourceEx->SetD3DManager(pManager));
+
+        return S_OK;
+    }
+
+    // IMFGetService methods
+    _Use_decl_annotations_
+    IFACEMETHODIMP HWMediaSource::GetService(
+            _In_ REFGUID guidService,
+            _In_ REFIID riid,
+            _Out_ LPVOID* ppvObject
+        )
+    {
+        winrt::slim_lock_guard lock(m_Lock);
+        wil::com_ptr_nothrow<IMFGetService> spGetService;
+
+        RETURN_IF_FAILED(_CheckShutdownRequiresLock());
+        if (SUCCEEDED(m_spDevSource->QueryInterface(IID_PPV_ARGS(&spGetService))))
+        {
+            RETURN_IF_FAILED(spGetService->GetService(guidService, riid, ppvObject));
+        }
+        else
+        {
+            return MF_E_UNSUPPORTED_SERVICE;
+        }
+
+        return S_OK;
+    }
+
+    // IKsControl methods
+    _Use_decl_annotations_
+    IFACEMETHODIMP HWMediaSource::KsProperty(
+            _In_reads_bytes_(ulPropertyLength) PKSPROPERTY pProperty,
+            _In_ ULONG ulPropertyLength,
+            _Inout_updates_to_(ulDataLength, *pBytesReturned) LPVOID pPropertyData,
+            _In_ ULONG ulDataLength,
+            _Out_ ULONG* pBytesReturned
+        )
+    {
+        // ERROR_SET_NOT_FOUND is the standard error code returned
+        // by the AV Stream driver framework when a miniport
+        // driver does not register a handler for a KS operation.
+        // We want to mimic the driver behavior here if we don't
+        // support controls.
+        winrt::slim_lock_guard lock(m_Lock);
+
+        RETURN_IF_FAILED(_CheckShutdownRequiresLock());
+        wil::com_ptr_nothrow<IKsControl> spKsControl;
+        if (SUCCEEDED(m_spDevSource->QueryInterface(IID_PPV_ARGS(&spKsControl))))
+        {
+            RETURN_IF_FAILED(spKsControl->KsProperty(pProperty, ulPropertyLength, pPropertyData, ulDataLength, pBytesReturned));
+        }
+        else
+        {
+            return HRESULT_FROM_WIN32(ERROR_SET_NOT_FOUND);
+        }
+
+        return S_OK;
+    }
+
+    _Use_decl_annotations_

```

Comment:
>Use_decl_annotations_ [](http://example.com/codeflow?start=5&length=21)

AFAIK, if you have this, then the method implementation shouldn't (or need not) have its own annotation.

Reply:
fix, remove _Use_decl_annotations_ keep annotation in cpp